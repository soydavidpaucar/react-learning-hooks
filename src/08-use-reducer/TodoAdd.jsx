import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ handleNewTodo }) => {
  const { description, handleChanges, handleResetForm } = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false
    };

    handleNewTodo(newTodo);
    handleResetForm();
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        placeholder="Add todo..."
        className="form-control"
        name="description"
        onChange={handleChanges}
        value={description}
      />

      <button type="submit" className="btn btn-outline-primary mt-2 btn-block">
        Submit
      </button>
    </form>
  );
};
