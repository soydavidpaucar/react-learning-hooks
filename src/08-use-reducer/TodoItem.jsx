export const TodoItem = ({ todo, deleteTodo, toggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between" onClick={() => toggleTodo(todo.id)}>
      <span className={`align-self-center${todo.done ? ' text-decoration-line-through' : ''}`}>{todo.description}</span>
      <button className="btn btn-danger" onClick={() => deleteTodo(todo.id)}>
        Delete
      </button>
    </li>
  );
};
