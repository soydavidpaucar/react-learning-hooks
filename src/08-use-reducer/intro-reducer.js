const initialState = [
  {
    id: 1,
    todo: 'Collect eggs',
    done: false
  }
];

/**
 * It returns the current state if the action type is not 'add todo', otherwise it returns a new array
 * with the current state and the new todo item
 * @param [state] - The current state of the reducer.
 * @param [action] - This is the action object that was dispatched.
 * @returns The state is being returned.
 */
const todoReducer = (state = initialState, action = {}) => {
  if (action.type == 'add todo') {
    return [...state, action.payload];
  }

  return state;
};

/* Initializing the todos variable with the initial state of the reducer. */
let todos = todoReducer();

/* Creating a new todo object. */
const newTodo = {
  id: 2,
  todo: 'Buy bread',
  done: false
};

/* Creating an action object. */
const addTodoAction = {
  type: 'add todo',
  payload: newTodo
};

/* Dispatching the action object to the reducer. */
todos = todoReducer(todos, addTodoAction);

console.log(todos);
