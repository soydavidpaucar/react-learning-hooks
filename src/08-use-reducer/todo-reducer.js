export const todoReducer = (initialState = [], action) => {
  switch (action.type) {
    case 'add':
      return [...initialState, action.payload];
    case 'delete':
      return initialState.filter((todo) => todo.id !== action.payload);
    case 'toggle':
      return initialState.map((todo) => {
        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done
          };
        }
        return todo;
      });
    default:
      return initialState;
  }
};
