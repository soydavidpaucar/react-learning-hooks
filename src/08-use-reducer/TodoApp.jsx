import { useEffect, useReducer } from 'react';

import { todoReducer } from './todo-reducer';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';

const initialState = [
  /* {
    id: new Date().getTime(),
    description: 'Learn React',
    done: false
  } */
];

/**
 * If there are todos in localStorage, parse them and return them, otherwise return an empty array
 * @returns An array of todos.
 */
const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, initialState, init);

  /* Saving the todos in localStorage. */
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleNewTodo = (newTodo) => {
    const action = {
      type: 'add',
      payload: newTodo
    };

    dispatch(action);
  };

  return (
    <>
      <h1>
        TodoApp {todos.length}, <small>Pending: 2</small>
      </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList todos={todos} />
        </div>
        <div className="col-5">
          <h4>Add Todo</h4>
          <hr />
          <TodoAdd handleNewTodo={handleNewTodo} />
        </div>
      </div>
    </>
  );
};
