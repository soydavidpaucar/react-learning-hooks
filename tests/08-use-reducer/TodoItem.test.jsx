import { render, screen, fireEvent } from '@testing-library/react';
import { beforeEach, describe, it, vi, expect } from 'vitest';
import { TodoItem } from '../../src/08-use-reducer/TodoItem';

describe('test on TodoItem', () => {
  const todo = {
    id: new Date().getTime(),
    description: 'Learn React',
    done: false
  };

  const deleteTodo = vi.fn();
  const toggleTodo = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = render(<TodoItem todo={todo} deleteTodo={deleteTodo} toggleTodo={toggleTodo} />);
    expect(container).toMatchSnapshot();
  });

  it('should display the incomplete todo item', () => {
    render(<TodoItem todo={todo} deleteTodo={deleteTodo} toggleTodo={toggleTodo} />);

    const liElement = screen.getByRole('listitem');
    const spanElement = document.querySelector('span').className;

    expect(liElement.className).toBe('list-group-item d-flex justify-content-between');
    expect(spanElement).toBe('align-self-center');
    expect(spanElement).not.toContain('text-decoration-line-through');
  });

  it('should display the complete todo item', () => {
    render(<TodoItem todo={{ ...todo, done: true }} deleteTodo={deleteTodo} toggleTodo={toggleTodo} />);

    const spanElement = document.querySelector('span').className;

    expect(spanElement).toBe('align-self-center text-decoration-line-through');
  });

  it('should call the toggleTodo function', () => {
    render(<TodoItem todo={todo} deleteTodo={deleteTodo} toggleTodo={toggleTodo} />);

    const liElement = screen.getByRole('listitem');

    fireEvent.click(liElement);

    expect(toggleTodo).toHaveBeenCalledTimes(1);
    expect(toggleTodo).toHaveBeenCalledWith(todo.id);
  });

  it('should call the deleteTodo function', () => {
    render(<TodoItem todo={todo} deleteTodo={deleteTodo} toggleTodo={toggleTodo} />);

    const buttonElement = screen.getByRole('button');

    fireEvent.click(buttonElement);

    expect(deleteTodo).toHaveBeenCalledTimes(1);
    expect(deleteTodo).toHaveBeenCalledWith(todo.id);
  });
});
