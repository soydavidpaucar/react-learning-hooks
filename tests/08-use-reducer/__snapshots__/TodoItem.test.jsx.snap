// Vitest Snapshot v1

exports[`test on TodoItem > should match snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      class="align-self-center"
    >
      Learn React
    </span>
    <button
      class="btn btn-danger"
    >
      Delete
    </button>
  </li>
</div>
`;
